src/llfs/track_fds.cpp

```diff
@@ -0,0 +1,188 @@
+//#=##=##=#==#=#==#===#+==#+==========+==+=+=+=+=+=++=+++=+++++=-++++=-+++++++++++
+//
+// Part of the LLFS Project, under Apache License v2.0.
+// See https://www.apache.org/licenses/LICENSE-2.0 for license information.
+// SPDX short identifier: Apache-2.0
+//
+//+++++++++++-+-+--+----- --- -- -  -  -   -
+
+#include <llfs/track_fds.hpp>
+//
+
+#include <llfs/logging.hpp>
+
+#include <batteries/env.hpp>
+#include <batteries/suppress.hpp>
+
+#include <filesystem>
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+
+#ifdef __linux__
+//
+#include <dirent.h>
+#include <sys/types.h>
+//
+#endif
+
+namespace llfs {
+
+namespace {
+
+const char* kVarName = "LLFS_TRACK_FDS";
+
+struct TrackFdState {
+  std::atomic<bool> enabled{batt::getenv_as<int>(kVarName).value_or(0) != 0};
+  std::mutex mutex;
+  std::unordered_map<int, std::unique_ptr<boost::stacktrace::stacktrace>> traces;
+
+  //+++++++++++-+-+--+----- --- -- -  -  -   -
+
+  static TrackFdState& instance() noexcept
+  {
+    // Intentionally leaked.
+    //
+    static TrackFdState* const instance_ = new TrackFdState{};
+
+    return *instance_;
+  }
+};
+
+}  //namespace
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+std::set<int> get_open_fds()
+{
+  std::set<int> result;
+  const char* dirpath = "/proc/self/fd";
+
+  //+++++++++++-+-+--+----- --- -- -  -  -   -
+#ifdef __linux__
+  DIR* p_dir = opendir(dirpath);
+
+  if (!p_dir) {
+    LLFS_LOG_ERROR() << "Could not open " << dirpath;
+  } else {
+    auto on_scope_exit = batt::finally([&] {
+      closedir(p_dir);
+    });
+
+    const int p_dir_fd = dirfd(p_dir);
+
+    isize name_max = pathconf(dirpath, _PC_NAME_MAX);
+    if (name_max == -1) { /* Limit not defined, or error */
+      name_max = 255;     /* Take a guess */
+    }
+    const usize len = offsetof(struct dirent, d_name) + name_max + 1;
+    void* const entry_buffer = malloc(len);
```

Comment:
`malloc` is C style. Do we need `malloc` here, or could we use `new` instead?

Reply:
This code snippet is derived from an old man page for `readdir_r`.  When I went to find that page, I re-read the doc for `readdir` (which isn't deprecated) more carefully, and realized it is a better thing to be using here.  Changed to use `readdir`, which eliminates the need for the malloc/free calls.